dnprogs/dntask/dntask.1
.TH DNTASK 1 "September 25 1998" "DECnet utilities"

.SH NAME
dntask \- Execute VMS command procedures
.SH SYNOPSIS
.B dntask
[options] command-procedure
.br
Options:
.br
[\-biVh] [\-t timeout]
.SH DESCRIPTION
.PP
dntask runs command procedures on a remote VMS system.
.br
It takes advantage of the ability of the TASK object in DECnet to
execute an arbitrary command procedure located in
the users' login directory. The remote command procedure should output
to SYS$NET rather than SYS$OUTPUT and (if interactive) read it's input
also from SYS$NET.
.br
The VMS equivalent of the (non-interactive version) command would be
.br

TYPE node::"TASK=command-procedure"

.br
For some example command procedures see the tasks directory of the source
distribution. 
.B show_system.com 
is a non-interactive task that simply displays
the output of the VMS "SHOW SYSTEM" command on standard output.
.B interactive.com
is an interactive task that allows you to enter DCL commands to be executed
on the host VMS system. Be careful which commands you enter because they will
expect input to come from the network connection, for instance programs that do
screen orientated input or output will almost certainly not work.
.br
Task names can be up to 16 characters in length because that's the limit on
DECnet object names.
.SH OPTIONS
.TP
.I "\-b"
Send the output in binary mode. By default the output from the DECnet task is
assumed to be records. This option sends the data "as is" so you can put
commands like BACKUP in the task and backup to your Linux box.
.TP
.I "\-i"
Interact with the command procedure. The command procedure must be written
to be interactive by reading from and writing to SYS$NET. Specifying 
.B \-i
for a non-interactive command procedure will cause 
.B dntask
to time-out waiting for input. Not specifying 
.B \-i
for an interactive command procedure will cause it to exit prematurely at
the VMS end.
.TP
.I "\-t timeout"
Specifies the timeout in seconds for interactive command procedures. If
no input has been received from either standard input or the VMS end
in this time then dntask will exit. The default is 60 seconds. If the value
0 is given then dntask will wait forever (or until you kill it).
.TP
.I "\-T connect timeout"
Specifies the maximum amount of time the command will wait to establish a connection
with the remote node. a 0 here will cause it to wait forever. The default is 60 seconds
.TP
.I \-h \-?
Displays help for using the command.
.TP
.I \-V
Show the version of the tools package that dntask comes from.
.SH NOTES
The command procedure that you write MUST ALWAYS write something to SYS$NET
or you will get a "connection refused" message. This is a limitation
with DECnet objects.
.br
eg. If you write a task to start a remote DECterm it would look something like
this:
.br

$ remnode=f$element(0, ":", "''f$trnlnm("sys$rem_node")'")
.br
$ set display/create/node='remnode'
.br
$ create/term/detach
.br
$ def/nolog sys$output sys$net
.br
$ write sys$output "DECterm started on ''remnode'"
.br
$ exit

The "write" command near the end is essential.

.SH EXAMPLES

  dntask 'myvax::show_system'

.br
  dndir \-i 'tramp"christine pjc123"::interactive'

.br
  dndir \-i 'tramp"christine \-"::interactive'

Specifying "-" in the password field will prompt for the password.

.SH SEE ALSO
.BR dntype "(1), " dndir "(1), " dncopy "(1), " dndel "(1)"


dnprogs/NEWS
dnprogs (2.61) unstable; urgency=low

  * Include libvaxdata/src/test.c in the distribution tarball
  * Fix password length in DAP connections. It was incorrectly set at 12 
    rather than 40
 
 -- Chrissie Caulfield <chrissie@debian.org>  Sun, 31 Mar 2013 11:18:32 +0000

dnprogs (2.60+x32) unreleased; urgency=low

  * Switch from using deprecated sysctl(2) to writing to /proc/sys.  This
    fixes a build failure on x32.

 -- Daniel Schepler <schepler@debian.org>  Tue, 12 Feb 2013 12:21:31 -0800

dnprogs (2.60) unstable; urgency=low

  * Fixed lots of Debian things that caused lintian problems and 
    upgraded policy version
    Closes: #673408
  * Fix some warnings that might be bugs
  * Incorporate NMU into main tree.

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 26 May 2012 09:54:41 +0100

dnprogs (2.59+nmu2) unstable; urgency=medium

  * NMU
  * dnprogs binaries are Linux-only, so state that. Move dnet-common back
    to binary-all. Closes: #665886

 -- Steve McIntyre <93sam@debian.org>  Mon, 21 May 2012 17:43:14 +0100

dnprogs (2.59+nmu1) unstable; urgency=low

  * NMU
  * Reduce libdnet's recommends on dnet-common to suggests again.
    Closes: #655740. Adapted patch from Jonathan Nieder.

 -- Steve McIntyre <93sam@debian.org>  Fri, 18 May 2012 13:29:53 +0100

dnprogs (2.59) unstable; urgency=low

  * Changed package dnet-common from arch all to any and
    updated dependecies. (Closes: #665886)

 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 27 Apr 2012 09:30:19 +0100

dnprogs (2.58) unstable; urgency=medium

  * Fixed lintian warnings:
    - Fixed changelog entry of last release to have correct line length.
    - Corrected debconf dependency in dnet-common.
    - Corrected versioned dependency in dnprogs on dnet-common.
    - Updated existing overrides to match new syntax.
  * Updated linkage of libdnet_daemon and libdap. (Closes: #646353)
  * Fixed compile warning in dndir when using -Werror=format-security
    (Closes: #646613)
  * Fixed typo in debian/dnet-common.config
  * Do not try to install files into packages on FreeBSD which are
    not build on FreeBSD.
  * Link dapfs with libpthread for new libfuse compatibility
  * Updated nl.po (Closes: #652349)
  * Added pl.po (Closes: #661894)
  * Added it.po (Closes: #663135)
  * Change module-init-tools dependency to kmod (Closes: #662684)

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 10 Mar 2012 10:32:08 +0000

dnprogs (2.57) unstable; urgency=medium

  [ Chrissie Caulfield ]
  * Completed removal of modutils configuration file code
  * Fix presence of CVS directories in tarball
  * updated 'clean' target

  [ Philipp Schafft ]
  * Use /dev/pts PTYs on 64 bit systems (fix makefile bug)
  * Updated translations Debconf templates:
    - German translation (Closes: #615268)
    - French translation (Closes: #615039)
    - Danish translation (Closes: #633984) (added)
    - Slovak translation (Closes: #639629) (added)
  * Look in yet another place for libcrypt (better fix than from NMUs)
  * Added a comment about the detecting of libcrypt. (See: #629664)
  * "ethernet" was capitalized in extended description (Closes: #636798)
  * Fixed some lintian warnings (mainly new rules targets).
  * dnet-common: Do not configure DECnet unless the user requested it.
    (Removed the bad workaround done by NMU) (See: #637179)
  * dnet-common: Added preinst script to clean up the broken config files
    from the last NMU. (Closes: #640861)
  * Upgraded libdnet's suggest to dnet-common to recommend again
    (Closes: #608807)

 -- Chrissie Caulfield <chrissie@debian.org>  Thu,  8 Sep 2011 16:05:47 +0200

dnprogs (2.56.1+nmu1) unstable; urgency=low

  * Non-maintainer upload.
  * libdnet: Only suggest dnet-common package (Closes: #608807)
  * dnet-common: Do not change MAC addresses unless user chooses to
    configure DECnet (Closes: #637179)
  * dnet-common: Restore ability to configure DECnet name and address
    through debconf
  * dnet-common: Remove obsolete modutils configuration file

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 25 Aug 2011 14:26:12 +0100

dnprogs (2.56.1) unstable; urgency=low

  * Non-maintainer upload.
  * Add patch from Ubuntu to fix FTBFS with libcrypt moved to the
    multiarch path.

 -- Aurelien Jarno <aurel32@debian.org>  Tue, 09 Aug 2011 11:58:06 +0200


dnprogs (2.56) unstable; urgency=low

  * Search for libgcrypt in 64bit library paths, too. (thanks to Kamil)
  * Package dnet-common now asks if network interfaces should be configrued
    for DECnet (Closes: #608807)
    Updated translations Debconf templates:
    - Russian translation (Closes: #610642)
    - Portuguese translation (Closes: #611034)
    - Swedish translation (Closes: #611296)
    - Czech translation (Closes: #611430)
    - Japanese translation (Closes: #611448)
    - Spanish translation (Closes: #611547)
  * Build against version 25 of FUSE
    Closes: #598831
  * Fixed some lintian warnings (typos and such).
  * Raised priority of package libdnet from extra to optional
    to comply policy manual section 2.5 "Priorities".

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 22 Feb 2011 10:38:23 +0000

dnprogs (2.55) unstable; urgency=low

  * Build against version 25 of FUSE
    Closes: #598831

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 5 Oct 2010 09:23:12 +0100

dnprogs (2.54) unstable; urgency=low

  * Don't build libvaxdata test program by default.
    Closes: #598752
 
 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 1 Oct 2010 20:44:09 +0100

dnprogs (2.53) unstable; urgency=low

  * Done some more porting to FreeBSD, corrected Build-Depends for hurd-i386
    Closes: #588185
  * Added new tool node(1) - Node name lookup tool
  * Added support to libdnet_daemon to read /etc/nodes.{allow,deny}
  * Fixed DoS bug in dnet_daemon()
    Closes: #592015 
  * Update libvaxdata to version 1.2, fixes double conversions

 -- Christine Caulfield <Christine.Caulfield@googlemail.com>  Thu, 19 Jul 2010 10:19:32 +0100


dnprogs/dntask/dntask.c
/******************************************************************************
    (c) 1998,2008   Christine Caulfield     christine.caulfield@googlemail.com
                    K.   Humborg            kenn@avalon.wombat.ie

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
 ******************************************************************************
 */
#include <sys/types.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <netdnet/dn.h>
#include <netdnet/dnetdb.h>
#include <fcntl.h>
#include <ctype.h>
#include "dn_endian.h"

struct	sockaddr_dn		sockaddr;
static  struct	nodeent		*np;
static  char			node[20];
static  int			sockfd;
static  int                     timeout = 60;
static  int                     connect_timeout = 60;

struct	sockaddr_dn		sockaddr;
struct	accessdata_dn		accessdata;
static  struct	nodeent        *binadr;
static  char    		filename[128];
static	unsigned char		buf[32760];
static  int			sockfd;
static  char                   *lasterror;
static  int                     binary_mode;

/* DECnet phase IV limits */
#define MAX_NODE      6
#define MAX_USER     12
#define MAX_PASSWORD 12
#define MAX_ACCOUNT  12
#ifndef FALSE
#define TRUE 1
#define FALSE 0
#endif

static int parse(char *);
static void usage(FILE *);
static char *connerror(int sockfd);
/*-------------------------------------------------------------------------*/

/*
 * Run an interactive command procedure. As we get input from either the
 * remote or local end we pass it on to the other.
 */
void be_interactive(void)
{
    fd_set         in;
    struct timeval tv;
    int            len;

    FD_ZERO(&in);
    FD_SET(sockfd, &in);
    FD_SET(STDIN_FILENO, &in);
    tv.tv_sec  = timeout;
    tv.tv_usec = 0;

    /* Loop for input */
    while ( select(sockfd+1, &in, NULL, NULL, (timeout==0?NULL:&tv)) > 0)
    {
	if (FD_ISSET(sockfd, &in)) // From VMS to us
	{
	    len = read(sockfd, buf, sizeof(buf));
	    if (len < 0)
	    {
		fprintf(stderr, "Read failed: %s\n", connerror(sockfd));
		return;
	    }
	    if (len == 0) /* EOF */
	    {
		return;
	    }

            if (binary_mode)
	    {
                write(STDOUT_FILENO, buf, len);
            }
	    else
	    {
	        buf[len] = '\0';
	        if (buf[len-1] == '\n') buf[len-1] = '\0';
		printf("%s\n", buf);
            }
	}
	if (FD_ISSET(STDIN_FILENO, &in)) // from us to VMS
	{
	    len = read(STDIN_FILENO, buf, sizeof(buf));
	    if (len < 0)
	    {
	        perror("reading from stdin");
		return;
	    }
	    if (len == 0) return; //EOF
	    if (buf[len-1] == '\n') buf[--len] = '\0';
	    write(sockfd, buf, len);
	}

	// Reset for another select
	FD_ZERO(&in);
	FD_SET(sockfd, &in);
	FD_SET(STDIN_FILENO, &in);
	tv.tv_sec  = timeout;
	tv.tv_usec = 0;
    }
    fprintf(stderr, "Time-out expired\n");
}

/*
 * Just dump whatever arrives onto stdout/
 */
void print_output(void)
{
    int len;

    while ( ((len = read(sockfd, buf, sizeof(buf)))) )
    {
        if (len == -1)
	{
	    if (errno != ENOTCONN)
		perror("Error reading from network");
	    else
		fprintf(stderr, "Read failed: %s\n", connerror(sockfd));
	    break;
        }

        if (binary_mode)
	{
            write(STDOUT_FILENO, buf, len);
        }
	else
	{
	    buf[len] = '\0';
	    if (buf[len-1] == '\n') buf[len-1] = '\0';
	    printf("%s\n", buf);
        }
    }
}

/*
 * Open the connection.
 */
int setup_link(void)
{
    struct timeval timeout = {connect_timeout,0};

    if ( (np=getnodebyname(node)) == NULL)
    {
	printf("Unknown node name %s\n",node);
	exit(0);
    }

/* Limit on the length of a DECnet object name */
    if (strlen(filename) > 16)
    {
        printf("task name must be less than 17 characters\n");
        exit(2);
    }

    if ((sockfd=socket(AF_DECnet,SOCK_SEQPACKET,DNPROTO_NSP)) == -1)
    {
	perror("socket");
	exit(-1);
    }

    // If no taskname was given then use a default.
    if (filename[0] == '\0')
    {
	strcpy(filename, "TASK");
    }

    if (setsockopt(sockfd, SOL_SOCKET, SO_SNDTIMEO, &timeout, sizeof(timeout)))
	    perror("Error setting snd timeout");

    // Provide access control and proxy information
    if (setsockopt(sockfd,DNPROTO_NSP,SO_CONACCESS,&accessdata,
		   sizeof(accessdata)) < 0)
    {
	perror("setsockopt");
	exit(-1);
    }

    /* Open up object number 0 with the name of the task */
    sockaddr.sdn_family   = AF_DECnet;

    /* Old kernel patches need flags to be 1 for a named object.
       The newer kernel patches define this as SDF_WILD so that's
       how we know
     */
    sockaddr.sdn_flags	  = 0x00;
    sockaddr.sdn_objnum	  = 0x00;
    memcpy(sockaddr.sdn_objname, filename, strlen(filename));
    sockaddr.sdn_objnamel = dn_htons(strlen(filename));
    memcpy(sockaddr.sdn_add.a_addr, np->n_addr,2);
    sockaddr.sdn_add.a_len = 2;


    if (connect(sockfd, (struct sockaddr *)&sockaddr, sizeof(sockaddr)) < 0)
    {
	fprintf(stderr, "Connect failed: %s\n", connerror(sockfd));
	exit(-1);
    }
    return TRUE;
}

/*
 * Start here
 */
int main(int argc, char *argv[])
{
    int opt;
    int interactive = 0;

    if (argc < 2)
    {
        usage(stdout);
        exit(1);
    }

    binary_mode = FALSE;

/* Get command-line options */
    opterr = 0;
    optind = 0;
    while ((opt=getopt(argc,argv,"?hVibt:T:")) != EOF)
    {
	switch(opt)
	{
	case 'h':
	    usage(stdout);
	    exit(1);

	case '?':
	    usage(stderr);
	    exit(1);

	case 'i':
	    interactive = 1;
	    break;

	case 't':
	    timeout = atoi(optarg);
	    break;

	case 'T':
	    connect_timeout = atoi(optarg);
	    break;

	case 'V':
	    printf("\ndntask from dntools version %s\n\n", VERSION);
	    exit(1);
	    break;

        case 'b':
            binary_mode = TRUE;
            break;
	}
    }
    if (!argv[optind])
    {
	usage(stderr);
	exit(1);
    }

    /* Parse the command into its parts */
    if (!parse(argv[optind]))
    {
	fprintf(stderr, "%s\n", lasterror);
	exit(2);
    }

    if (!setup_link()) return 2;

/* Do the real work */
    if (!interactive)
	print_output();
    else
	be_interactive();

    close(sockfd);
    return 0;
}


/*-------------------------------------------------------------------------*/
/* Parse the filename into its component parts */
static int parse(char *fname)
{
    enum  {NODE, USER, PASSWORD, ACCOUNT, NAME, FINISHED} state;
    int   n0=0;
    int   n1=0;
    char  sep, term;
    char *local_user;

    state = NODE; /* Node is mandatory */

    while (state != FINISHED)
    {
	switch (state)
	{
	case NODE:
            if (fname[n0] != ':' && fname[n0] != '\"' &&
                fname[n0] != '\'' && fname[n0] != '/')
            {
                if (n1 >= MAX_NODE ||
                    fname[n0] == ' ' || fname[n0] == '\n')
                {
                    lasterror = (char *)"File name parse error";
                    return FALSE;
                }
                node[n1++] = fname[n0++];
            }
            else
            {
                node[n1] = '\0';
                n1 = 0;
                switch (fname[n0])
                {
                case '\"':
                case '\'':
                    sep = ' ';
                    term = fname[n0++];
                    state = USER;
                    break;

                case '/':
                    sep = '/';
                    term = 0;
                    n0++;
                    state = USER;
                    break;

                default:
                    n0 += 2;
                    state = NAME;
                    break;
                }
            }
            break;

	case USER:
            if (fname[n0] != sep && fname[n0] != (term ? term : ':'))
            {
                if (n1 >= MAX_USER)
                {
                    lasterror = (char *)"File name parse error";
                    return FALSE;
                }
                accessdata.acc_user[n1++] = fname[n0++];
            }
            else
            {
                accessdata.acc_user[n1] = '\0';
                n1 = 0;
                if (fname[n0] == sep)
                {
                    state = PASSWORD;
                    n0++;
                }
                else
                {
                    state = NAME;
                    n0 += term ? 3 : 2;
                }
            }
            break;

	case PASSWORD:
            if (fname[n0] != sep && fname[n0] != (term ? term : ':'))
            {
                if (n1 >= MAX_PASSWORD)
                {
                    lasterror = (char *)"File name parse error";
                    return FALSE;
                }
                accessdata.acc_pass[n1++] = fname[n0++];
            }
            else
            {
                accessdata.acc_pass[n1] = '\0';
                n1 = 0;
                if (fname[n0] == sep)
                {
                    state = ACCOUNT;
                    n0++;
                }
                else
                {
                    state = NAME;
                    n0 += term ? 3 : 2;
                }
            }
            break;

	case ACCOUNT:
            if (fname[n0] != sep && fname[n0] != (term ? term : ':'))
            {
                if (n1 >= MAX_ACCOUNT)
                {
                    lasterror = (char *)"File name parse error";
                    return FALSE;
                }
                accessdata.acc_acc[n1++] = fname[n0++];
            }
            else
            {
                accessdata.acc_acc[n1] = '\0';
                n1 = 0;
                state = NAME;
                n0 += term ? 3 : 2;
            }
            break;

	case NAME:
	    strcpy(filename, fname+n0);
	    state = FINISHED;
	    break;

	case FINISHED: // To keep the compiler happy
	    break;
	} /* switch */
    }

/* tail end validation */
    binadr = getnodebyname(node);
    if (!binadr)
    {
	lasterror = "Unknown or invalid node name ";
	return FALSE;
    }

    // Try very hard to get the local username
    local_user = cuserid(NULL);
    if (!local_user || local_user == (char *)0xffffffff)
      local_user = getenv("LOGNAME");
    if (!local_user) local_user = getenv("USER");
    if (local_user)
    {
        int i;

	strcpy((char *)accessdata.acc_acc, local_user);
	accessdata.acc_accl = strlen((char *)accessdata.acc_acc);
        for (i=0; i<accessdata.acc_accl; i++)
            accessdata.acc_acc[i] = toupper(accessdata.acc_acc[i]);
    }
    else
        accessdata.acc_acc[0] = '\0';



    /* If the password is "-" and fd 0 is a tty then
       prompt for a password */
    if (accessdata.acc_pass[0] == '-' && accessdata.acc_pass[1] == '\0' && isatty(0))
    {
	char *password = getpass("Password: ");
	if (password == NULL || strlen(password) > (unsigned int)MAX_PASSWORD)
	{
	    lasterror = "Password input cancelled";
	    return FALSE;
	}
	strcpy((char *)accessdata.acc_pass, password);
    }


    /* Complete the accessdata structure */
    accessdata.acc_userl = strlen((char *)accessdata.acc_user);
    accessdata.acc_passl = strlen((char *)accessdata.acc_pass);
    accessdata.acc_accl  = strlen((char *)accessdata.acc_acc);
    return TRUE;
}

static void usage(FILE *f)
{
    fprintf(f, "\nUSAGE: dntask [OPTIONS] 'node\"user password\"::commandprocedure'\n\n");
    fprintf(f, "NOTE: The VMS filename really should be in single quotes to\n");
    fprintf(f, "      protect it from the shell\n");

    fprintf(f, "\nOptions:\n");
    fprintf(f, "  -i           Interact with the command procedure\n");
    fprintf(f, "  -t           Timeout (in seconds) for interactive command procedure input\n");
    fprintf(f, "  -T <secs>    Connect timeout (default 60 seconds)\n");
    fprintf(f, "  -b           Treat received data as binary data\n");
    fprintf(f, "  -? -h        display this help message\n");
    fprintf(f, "  -V           show version number\n");

    fprintf(f, "\nExamples:\n\n");
    fprintf(f, " dntask 'myvax::'  - defaults to TASK.COM and proxy username\n");
    fprintf(f, " dntask -i 'clustr\"christine thecats\"::do_dcl.com'\n");
    fprintf(f, "\n");
}

// Return the text of a connection error
static char *connerror(int sockfd)
{
#ifdef DSO_CONDATA
    struct optdata_dn optdata;
    unsigned int len = sizeof(optdata);
    char *msg;

    if (errno == ETIMEDOUT || getsockopt(sockfd, DNPROTO_NSP, DSO_DISDATA,
		   &optdata, &len) == -1)
    {
	return strerror(errno);
    }

    // Turn the rejection reason into text
    switch (optdata.opt_status)
    {
    case DNSTAT_REJECTED: msg="Rejected by object";
	break;
    case DNSTAT_RESOURCES: msg="No resources available";
	break;
    case DNSTAT_NODENAME: msg="Unrecognised node name";
	break;
    case DNSTAT_LOCNODESHUT: msg="Local Node is shut down";
	break;
    case DNSTAT_OBJECT: msg="Unrecognised object";
	break;
    case DNSTAT_OBJNAMEFORMAT: msg="Invalid object name format";
	break;
    case DNSTAT_TOOBUSY: msg="Object too busy";
	break;
    case DNSTAT_NODENAMEFORMAT: msg="Invalid node name format";
	break;
    case DNSTAT_REMNODESHUT: msg="Remote Node is shut down";
	break;
    case DNSTAT_ACCCONTROL: msg="Login information invalid at remote node";
	break;
    case DNSTAT_NORESPONSE: msg="No response from object";
	break;
    case DNSTAT_NODEUNREACH: msg="Node Unreachable";
	break;
    case DNSTAT_MANAGEMENT: msg="Abort by management/third party";
	break;
    case DNSTAT_ABORTOBJECT: msg="Remote object aborted the link";
	break;
    case DNSTAT_NODERESOURCES: msg="Node does not have sufficient resources for a new link";
	break;
    case DNSTAT_OBJRESOURCES: msg="Object does not have sufficient resources for a new link";
	break;
    case DNSTAT_BADACCOUNT: msg="The Account field in unacceptable";
	break;
    case DNSTAT_TOOLONG: msg="A field in the access control message was too long";
	break;
    default: msg=strerror(errno);
	break;
    }
    return msg;
#else
    return strerror(errno);

#endif
}
